### PR TITLE
You can now drag things up and down ladders

### PR DESCRIPTION
**Pull request by imtakingabreakdontatme**
Things you are pulling will go with you

Also added a subtype of ladder that autoconnects to ladders that share the same X and Y

Fixed ladder icons not updating when they lose a connection

### CODE REVIEW

File: code/game/objects/structures/ladders.dm [hunk 1]
```diff
@@ -50,31 +50,27 @@
...
+	user.forceMove(get_turf(ladder))
+	if(user.pulling)
+		user.pulling.forceMove(get_turf(ladder))
 
 /obj/structure/ladder/proc/use(mob/user,is_ghost=0)
 	if(up && down)
 		switch( alert("Go up or down the ladder?", "Ladder", "Up", "Down", "Cancel") )
 			if("Up")
-				go_up(user,is_ghost)
+				travel(1, user, is_ghost, up)
```
Comment (ChangelingRain) [inline code/game/objects/structures/ladders.dm]: TRUE

File: code/game/objects/structures/ladders.dm [hunk 2]
```diff
@@ -50,31 +50,27 @@
...
 			if("Down")
-				go_down(user,is_ghost)
+				travel(0, user, is_ghost, down)
```
Comment (ChangelingRain) [inline code/game/objects/structures/ladders.dm]: FALSE

File: code/game/objects/structures/ladders.dm [hunk 3]
```diff
@@ -50,31 +50,27 @@
...
 			if("Cancel")
 				return
 	else if(up)
-		go_up(user,is_ghost)
+		travel(1, user, is_ghost, up)
```
Comment (ChangelingRain) [inline code/game/objects/structures/ladders.dm]: TRUE

File: code/game/objects/structures/ladders.dm [hunk 4]
```diff
@@ -50,31 +50,27 @@
...
 	else if(down)
-		go_down(user,is_ghost)
+		travel(0, user,is_ghost, down)
```
Comment (ChangelingRain) [inline code/game/objects/structures/ladders.dm]: FALSE

File: code/game/objects/structures/ladders.dm [hunk 5]
```diff
@@ -50,31 +50,27 @@
...
-
-/obj/structure/ladder/proc/go_down(mob/user,is_ghost)
-	if(!is_ghost)
-		show_fluff_message(0,user)
-		down.add_fingerprint(user)
-	user.loc = get_turf(down)
+		show_fluff_message(going_up,user)
+		ladder.add_fingerprint(user)
+	user.forceMove(get_turf(ladder))
+	if(user.pulling)
```
Comment (ChangelingRain) [inline code/game/objects/structures/ladders.dm]: you need to do this before moving the user, and you should also reestablish the pull after both move

**Comment by FantasticFwoosh** [comment]
We're going all 'baystation' over here with ladder shenanigans, i guess it would need to be done in the case of ever making that mining under layer (unless you just worked on your elevators more)

**Comment by Tokiko1** [comment]
If you want to use these ladders for mapping, don't forget to make them anchored.

**Comment by RemieRichards** [comment]
I disagree.
You should limit the weight ability to do this, at the very least.

You should have to choose between quick ladders and no pulling, or slow stairs and pulling your gear.

**Comment by FantasticFwoosh** [comment]
Why not just choose a elevator option instead? because stairs would be pointlessly the same action with minimal variation.

**Comment by RemieRichards** [comment]
We have elevators already (they're just shuttles, same as with other codebases), I didn't mention them because there's no real need, an elevator (if working) will always be the best option.

**Comment by KorPhaeron** [comment]
Could just add a channeling period for taking things up ladders, would be a pain to fit both ladders and stairs everywhere on the map

**Comment by RemieRichards** [comment]
I'm not saying have both.
I'm saying put care and attention as to what could exactly go down in that area, and choose the type of traversal which will make those events play off even better.

**Comment by FantasticFwoosh** [comment]
No but that creates a obvious problem @RemieRichards (i am aware about elevators, i looked at the PR while i was skimming over this) in which mappers may intentionally leave out a elevator/stairs option and force players to disassemble things piecemeal as a alternative or out of lazyness not knowing how to construct said elevator or if the stairs does not fit the aesthetic.

- On that note of dragging things, dropping as opposed to dragging mobs down stairs should most definitely hurt and be a option on harmful intent.
____________________________________________

Time to make a infinite looping stair z level to push miners who pick the wrong path down dealing brute damage on every step.

**Comment by Tacolizard** [comment]
just make ladder pull speed based on the size of the object, we already have vars for that.

**Review by Cyberboss — CHANGES_REQUESTED** [review]
Do Joan's stuff

**Comment by KorPhaeron** [comment]
Stale, will port jade mansions ladders in another PR

**Comment by KorPhaeron** [comment]
@Cyberboss @ChangelingRain 

I did the thing

File: code/game/objects/structures/ladders.dm [hunk 6]
```diff
@@ -50,31 +57,34 @@
...
+		show_fluff_message(going_up,user)
+		ladder.add_fingerprint(user)
+
+	var/atom/movable/AM
+	if(user.pulling)
+		AM = user.pulling
+		user.pulling.forceMove(get_turf(ladder))
+	user.forceMove(get_turf(ladder))
+	if(AM)
+		user.pulling = AM
```
Comment (AnturK) [inline code/game/objects/structures/ladders.dm]: I'm pretty sure there's a helper to start pulling, you should use that.

File: code/game/objects/structures/ladders.dm [hunk 7]
```diff
@@ -19,12 +20,18 @@
 	return INITIALIZE_HINT_LATELOAD
 
 /obj/structure/ladder/Destroy()
+	if(up)
+		up.down = null
```
Comment (AnturK) [inline code/game/objects/structures/ladders.dm]: You could technically have ladders that go down to different place.

**Review by AnturK — CHANGES_REQUESTED** [review]

File: code/game/objects/structures/ladders.dm (hunk 7 above)
Comment (KorPhaeron) [reply code/game/objects/structures/ladders.dm]: That'd be pretty terrible but good point

**Review by AnturK — APPROVED** [review]

**Pull request merged**